core/trino-main/src/test/java/io/trino/sql/planner/optimizations/TestAddExchangesPlans.java

```diff
@@ -520,6 +525,202 @@ public void testJoinBuildSideLocalExchange()
                                                         tableScan("region", ImmutableMap.of("regionkey", "regionkey"))))))));
     }
 
+    @Test
+    public void testAggregateIsExactlyPartitioned()
```

Comment:
can we have negative tests too? for queries like that but with `USE_EXACT_PARTITIONING` == `false`. Currently other tests already defined here do not distinguish if `USE_EXACT_PARTITIONING` is true of false

Reply:
Added a negative test `testJoinNotExactlyPartitioned`.

It's hard for add negative tests for other query shapes, because other query shapes' query plans are linear lists, the only difference is the number of exchanges. To write such negative tests you need to construct the whole query plan, which is really time-consuming.